### ISSUE TITLE
Stall of `DLAMC1` using IFX IntelLLVM 2024.2.1 in release mode.

### ISSUE THREAD

**Issue by V-Chavarrias** [Reference-LAPACK/lapack#1054]
**Description**

Subroutine `DLAMC1` stalls when using IFX Fortran compiler in release mode. 

The stalling occurs here:

```
   10    CONTINUE
         IF( C.EQ.ONE ) THEN
            A = 2*A
            C = DLAMC3( A, ONE )
            C = DLAMC3( C, -A )
            GO TO 10
         END IF
```

The idea is to find the largest possible number on the used combination of machine, compiler, optimizations flags, etcetera. Two double precission floatings `A` and `C` are initialized as 1. `A` is doubled and `C=A+1`. Then, `C=C-A`. If the number can still be represented in the machine, `C=1` and the loop continues. At some point, `A` is so large that, when adding 1 it remains being `A`. Hence, `C=A`. It is the maximum it can be. Then, when subtracting `A-C=0`. Now, 0 is different than 1 and the loop stops. The largest number has been found.

A set of write-to-file statements are added in debug mode to see the correct behaviour of the loop.

In debug mode:

```
A   0.1000000000000000E+01
2A  0.2000000000000000E+01
C   0.3000000000000000E+01
C2  0.1000000000000000E+01
A   0.2000000000000000E+01
2A  0.4000000000000000E+01
C   0.5000000000000000E+01
C2  0.1000000000000000E+01

...

A   0.2251799813685248E+16
2A  0.4503599627370496E+16
C   0.4503599627370497E+16
C2  0.1000000000000000E+01
A   0.4503599627370496E+16
2A  0.9007199254740992E+16
C   0.9007199254740992E+16
C2  0.0000000000000000E+00

```

```
In release mode with IFX, the following is observed:

A   0.2814749767106560E+15
2A  0.5629499534213120E+15
C   0.5629499534213130E+15
C2  0.1000000000000000E+01
A   0.5629499534213120E+15
2A  0.1125899906842624E+16
C   0.1125899906842625E+16
C2  0.1000000000000000E+01

...

A   0.2251799813685248E+16
2A  0.4503599627370496E+16
C   0.4503599627370497E+16
C2  0.1000000000000000E+01
A   0.4503599627370496E+16
2A  0.9007199254740992E+16
C   0.9007199254740992E+16
C2  0.1000000000000000E+01

...

A   0.4494232837155790E+308
2A  0.8988465674311580E+308
C   0.8988465674311580E+308
C2  0.1000000000000000E+01
A   0.8988465674311580E+308
2A                 Infinity
C                 Infinity
C2  0.1000000000000000E+01
A                 Infinity
2A                 Infinity
C                 Infinity
C2  0.1000000000000000E+01

```

We see that, when C=0.9007199254740992E+16, it is the same as 2A, but still when subtracting them we obtain 1, the value that we have "added". The compiler is so optimized that it does not do the subtraction. It already "knows" that the answer is 1. Hence, the loop runs infinitely.

I have found this in a version of LAPACK compiled in Delft3D. The version of LAPACK is actually not the latest, but the exact same loop is in the latest version. Hence the same problem must occur. 

This problem does not have an easy solution. The logic behind the loop is broken if the compiler is "so smart". Also, the need to find `eps` using such an algorithm may not be needed nowadays. 

**Checklist**

- [ ] I've included a minimal example to reproduce the issue
- [ ] I'd be willing to make a PR to solve this issue

**Comment by langou** [comment]
Thanks @V-Chavarrias for the nice explanation on how DLAMC1 works.

I agree that @martin-frbg suggestion ( use '-fp-model strict') should be tried.

Another idea could be to remove 
      DOUBLE PRECISION FUNCTION DLAMC3( A, B )
out of dlamchf77.f so as to compile DLAMC1 and DLAMC3 "separately". So do not give both functions in the same file to the compiler, but give the two functions in different files and compile each file separately. In other words, create a dlamc3.f file. Compile it. And remove the function from dlamchf77.f. Something like that. The hope is to prevent this kind of aggressive compiler optimization.

Note that the use of DLAMC1 (where the infinite loop occurs) is less important these days since we rely more on IEEE arithmetic standard and the FORTRAN INTRINSICS to compute the output of DLAMCH. DLAMC1 was used in a previous version of DLAMCH and this previous version of DLAMCH should not really be used if the more modern DLAMCH works.

The current DLAMCH is at: 
https://github.com/Reference-LAPACK/lapack/blob/master/INSTALL/dlamch.f

The older DLAMCH (where DLAMC1 is) is at:
https://github.com/Reference-LAPACK/lapack/blob/master/INSTALL/dlamchf77.f

Julien.

**Comment by martin-frbg** [comment]
CheckLapackCompilerFlags.cmake already adds the fp-model option for Intel Fortran, but it only does this on "UNIX" and  I have not checked if the build system for Delft3D uses the provided cmake files for building LAPACK at all.

**Comment by V-Chavarrias** [comment]
I analyzed the problem in Windows, but the problem is also in Linux (it was identified in a CI/CD pipeline on Linux).

I was using a version of the old `DLAMCH`. Maybe with the new one is not a problem. I should try.  It is not easy for me to just change the version because of how it is integrated in the program, but I will try. 

One comment. I am not sure this is a compiler bug nor a bug in `LAPACK`. It seems to me that the algorithm is correct, but it was not designed for such an advanced compiler. It is just outdated and should not be used. To overcome the issue, I just used IEEE `eps` at a higher lever in the program and simply did not call `DLAMCH`. 

Maybe there should be a warning or it should we in a list of known issues such that the next person finds this when googling the problem.
